partitura/musicanalysis/meter.py

```diff
@@ -0,0 +1,310 @@
+            self.agents = []
+            for i in range(len(agents_all)):
+                for j in range(i+1, len(agents_all)):
+                    if duplicate[i] > 0 or duplicate[j] > 0:
+                        continue
+
+                    if abs(agents_all[i].beat_interval - agents_all[j].beat_interval) < 0.01 \
+                            and abs(agents_all[i].lastBeat() - agents_all[j].lastBeat()) < 0.02:
+                        if agents_all[i].score > agents_all[j].score:
+                            duplicate[j] += 1
+                        else:
+                            duplicate[i] += 1
+                            break
+
+            self.agents = sorted(np.asarray(agents_all)[(duplicate < 1)].tolist(
+            ), key=lambda x: x.score, reverse=True)[:MAX_AGENTS]
+
+        self.agents = sorted(self.agents, key=lambda x: x.score, reverse=True)
+
+
+class Cluster():
+    
+    def __init__(self, ioi) -> None:
+        self.iois = np.zeros(0)
+        self.score = 0
+        self.interval = 0
+        self.addIoi(ioi)
+
+    def getK(self, ioi):
+        diff = abs(self.interval-ioi)
+        if diff < CLUSTER_WIDTH:
+            return diff
+        return False
+
+    def addIoi(self, ioi):
+        self.iois = np.append(self.iois, ioi)
+        self.interval = np.sum(self.iois)/len(self.iois)
+
+    @staticmethod
+    def relationship_factor(d):
+        if 1 <= d and d <= 4:
+            return 6-d
+        elif 5 <= d and d <= 8:
+            return 1
+        return 0
+
+class Agent():
+
+    def __init__(self) -> None:
+        self.beat_interval = 0
+        self.prediction = 0
+        self.history = []
+        self.score = 0
+
+    def lastBeat(self):
+        i = len(self.history)-1
+        while i > 0 and self.history[i][1] == 0:
+            i-=1
+        return self.history[i][0]
+
+    def getTempo(self):
+        return 60.0 * (len(self.history)-1) / (self.history[-1][0]-self.history[0][0])
+    
+    def getTimeSignatureNum(self):
+        possibleNums = [2, 3, 4, 6, 9, 12, 24]
+        bestVal = {num:0 for num in possibleNums}
+        salience = list(zip(*self.history))[1]
+        sumSalience = sum(salience)
+        f = 1.005
+        for num in possibleNums:
+            for startIdx in range(num):    
+                dbs = len(salience[startIdx::num])
+                if dbs > 1:
+                    downbeatSalience = sum(salience[startIdx::num])/dbs
+                    sumSalience = sum(salience[:(dbs-1)*num]) 
+                    otherSalience = (sumSalience-downbeatSalience*dbs)/((num-1)*(dbs-1))
+                else:
+                    downbeatSalience = 0
+                    otherSalience = 1
+                    
+                ratio = downbeatSalience/otherSalience
+                bestVal[num] = max(bestVal[num], ratio)
+
+        bestNum = max(bestVal, key=bestVal.get)    
+            
+        return bestNum
+
+
+def estimate_time(note_info):
```

Comment:
Should we make sure that `estimate_time` also accepts a Score object?

Reply:
like all these functions it calls ```ensure_note_array``` on the argument passed. Ensure note array should be updated to work with scores, performances (if it isn't already)

Reply:
I checked, ensure note array  works with score objects